### PR TITLE
Login not working starting with XWiki 15.9 #41

### PR DESCRIPTION
**Pull request by oanalavinia**
* get the encryption key automatically generated and stored in the permanent directory, since starting with newer xwiki versions, there is no default value for the encryption key in xwiki.cfg
* duplicate code from the PermanentDirectory xwiki class, which is not available in older xwiki versions, in order to access the encryption key

### CODE REVIEW

File: api/src/main/java/com/xwiki/identityoauth/internal/PermanentConfigurationSource.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    private Configuration loadConfiguration() throws InitializationException
+    {
+        File permanentDirectory = this.environment.getPermanentDirectory();
+        File permanentConfiguration = new File(permanentDirectory, XWIKI_PROPERTIES_FILE).getAbsoluteFile();
+
+        // If it does not exist, create it.
+        if (!permanentConfiguration.exists()) {
+            try {
+                FileUtils.write(permanentConfiguration, "", StandardCharsets.UTF_8);
```
Comment (trrenty) [inline api/src/main/java/com/xwiki/identityoauth/internal/PermanentConfigurationSource.java:80]: Do you really need to create the file if it does not exist? From what I understand, if the wiki instance is greater than 15.9, the file should exist and contain the security params, right?

File: api/src/main/java/com/xwiki/identityoauth/internal/CookieAuthenticationPersistence.java [hunk 2]
```diff
@@ -183,14 +183,29 @@ void setUserId(String userUid)
...
+        String key = xwikiCfg.getProperty(ENCRYPTION_KEY_PROPERTY);
+        // In case the property was not defined, fall back on the XWiki encryption key property value.
+        if (key == null) {
+            key = xwikiCfg.getProperty(XWIKI_ENCRYPTION_KEY_PROPERTY);
+        }
+        // Starting with XWIKI-542:The cookie encryption keys should be randomly generated, when the encryption key is
+        // not declared, it's value is automatically generated and stored in the permanent directory, instead of using
+        // default values as before.
+        if (key == null) {
+            key = this.permanentConfiguration.getProperty(XWIKI_ENCRYPTION_KEY_PROPERTY, String.class);
```
Comment (mflorea) [inline api/src/main/java/com/xwiki/identityoauth/internal/CookieAuthenticationPersistence.java:197]: You could add a ``getPermanentConfiguration()`` method that uses the component manager to look for the (standard) ``permanent`` ``ConfigurationSource``, falling back on your own ``"identity/permanent"`` implementation. The advantage with this approach is that on newer XWiki versions the standard implementation is used, which may receive bug fixes or improvements in time.

File: api/src/main/java/com/xwiki/identityoauth/internal/PermanentConfigurationSource.java (hunk 1 above)
Comment (oanalavinia) [reply api/src/main/java/com/xwiki/identityoauth/internal/PermanentConfigurationSource.java:80]: This code is copied from https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-configuration/xwiki-platform-configuration-default/src/main/java/org/xwiki/configuration/internal/PermanentConfigurationSource.java , so this is the actual code that creates the file. So for older versions, the file won't exist and we would need to create it, while on newer versions the check is needed

**Comment by oanalavinia** [comment]
I deleted my `identity/permanent` implementation since it's role was just to give access to the key value stored in newer versions of xwiki, not to backport the feature to older versions that don't have this implemented yet.
And this was easily done using the component manager. Thanks both of you for the suggestions!

**Pull request merged**